File: challenges/digital_root.rb
# Write a method, digital_root(num). It should sum the digits of a
# positive integer. If it is greater than or equal to 10, sum the
# digits of the resulting number. Keep repeating until there is only
# one digit in the result, called the "digital root". Do not use
# string conversion within your method.

def digital_root(num)
  while num >= 10
    num = digital_root_step(num)
  end

  num
end

def digital_root_step(num)
  root = 0
  while num > 0
    root += (num % 10)

    num /= 10
  end

  root
end

#recursive solution

def digital_root_recur num
  return num if num < 10
  digital_root_recur((num % 10) + (num / 10))
end

File: challenges/digital_root.js
// # Write a method, digital_root(num). It should sum the digits of a
// # positive integer. If it is greater than or equal to 10, sum the
// # digits of the resulting number. Keep repeating until there is only
// # one digit in the result, called the "digital root". Do not use
// # string conversion within your method.

function digitalRoot(num) {
  while (num >= 10) {
    num = rootStep(num);
  }

  return num;
}

function rootStep(num) {
  let sum = 0;

  while (num !== 0) {
    let digit = num % 10;
    sum += digit;
    num = Math.floor(num / 10);
  }

  return sum;
}
